### PR TITLE
[Validation] fix setting all model_parts to eval mode

### PR DESCRIPTION
**Pull request by wesleytruong**
Sets each of `model_parts` to eval mode in `Validator` class to support PP.

### CODE REVIEW

File: torchtitan/components/validate.py [hunk 1]
```diff
@@ -82,8 +82,9 @@ def validate(
         step: int,
     ) -> None:
         # Set model to eval mode
+        for model in model_parts:
+            model.eval()
         model = model_parts[0]
```
Comment (tianyu-l) [inline torchtitan/components/validate.py]: why keeping this? I think we only need this in non-PP case.

File: torchtitan/components/validate.py [hunk 2]
```diff
@@ -174,7 +175,8 @@ def validate(
                 module.reshard()
```
Comment (tianyu-l) [inline torchtitan/components/validate.py]: After you switch the order of checkpoint and validate, I think this reshard part is not necessary and creates extra overhead -- we are doing validate (reshard) -> next train (unshard), which we could've avoided?

Could you check if it works well when checkpoint and validate happens on the same step? If so we can remove this code.

Comment (wesleytruong) [reply torchtitan/components/validate.py]: Sure, in my test it looks like removing this reshard doesn't affect memory usage or loss so I think it should be fine to remove.

| Reshard | No Reshard |
| --- | --- |
| Screenshot 2025-08-20 at 3 10 20 PM | Screenshot 2025-08-20 at 3 36 45 PM |

File: torchtitan/train.py [hunk 3]
```diff
@@ -309,7 +309,11 @@ def __init__(self, job_config: JobConfig):
             checkpoint_config=job_config.checkpoint,
             sd_adapter=(
                 self.train_spec.state_dict_adapter(
-                    model_args, job_config.model.hf_assets_path
+                    model_args,
+                    job_config.model.hf_assets_path
+                    if job_config.checkpoint.enable_checkpoint
```
Comment (wwwjn) [inline torchtitan/train.py]: Why we need this change here?  hf_assets_path can be built during loading HF weights as well, even if it's not used, and we could leave the checkpointing related logic in `checkpoint.py`

Comment (wesleytruong) [reply torchtitan/train.py]: Sorry you're right, I was trying to do some clean up in order to suppress this warning if the user is not intending to save in HF since in that case we wouldn't need a `model.safetensor.index.json`, but this would interfere with if we want to load from HF from the `hf_assets_path`. One other idea to suppress this would be to move this error to the checkpointer, but then it would make this warning difficult to overload such as we want to may want to do in Flux. https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/model/state_dict_adapter.py#L53-L57

Comment (wwwjn) [reply torchtitan/train.py]: I think it's ok to show the warning to user. When do we want to suppress the warning? I could only think if a model has only one safetensor file, which out the model.safetensor.index.json. 

If a model checkpoint has multiple safetensor file, then both saving and loading needs to check the `model.safetensor.index.json` file exists.

Comment (tianyu-l) [reply torchtitan/train.py]: > When do we want to suppress the warning?

E.g. when checkpointing is not enabled at all.

> If a model checkpoint has multiple safetensor file, then both saving and loading needs to check the `model.safetensor.index.json` file exists.

I believe only saving requires it. For loading it should be optional, but I'm not sure if it helps load faster. cc @ankitageorge to confirm.

Comment (ankitageorge) [reply torchtitan/train.py]: Ya only saving requires it

File: torchtitan/train.py [hunk 4]
```diff
@@ -309,7 +309,11 @@ def __init__(self, job_config: JobConfig):
             checkpoint_config=job_config.checkpoint,
             sd_adapter=(
                 self.train_spec.state_dict_adapter(
-                    model_args, job_config.model.hf_assets_path
+                    model_args,
+                    job_config.model.hf_assets_path
+                    if job_config.checkpoint.enable
+                    and job_config.checkpoint.last_save_in_hf
+                    else None,
```
Comment (tianyu-l) [inline torchtitan/train.py]: hmmm even for load you need the path? https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/checkpoint.py#L543

I think there are two ways:
1. always leave the warning there as is
2. only pass the `hf_assets_path` in when `checkpoint.enable=True`

I feel since 2 is not clean enough (as in we don't differentiate save vs. load), so I'm OK with 1.
cc @wwwjn for your opinion.

Comment (wwwjn) [reply torchtitan/train.py]: I'm OK with 1 as well. We can leave the warning and let user aware of this

File: torchtitan/config/job_config.py [hunk 5]
```diff
@@ -398,13 +398,13 @@ class Parallelism:
 
 @dataclass
 class Checkpoint:
-    enable_checkpoint: bool = False
+    enable: bool = False
```
Comment (fegin) [inline torchtitan/config/job_config.py:401]: nice, I was thinking to do this as well.

File: torchtitan/train.py [hunk 6]
```diff
@@ -309,7 +309,13 @@ def __init__(self, job_config: JobConfig):
             checkpoint_config=job_config.checkpoint,
             sd_adapter=(
                 self.train_spec.state_dict_adapter(
-                    model_args, job_config.model.hf_assets_path
+                    model_args,
+                    (
+                        job_config.model.hf_assets_path
+                        if job_config.checkpoint.enable
```
Comment (fegin) [inline torchtitan/train.py]: Could you please move using checkpoint information out of the trainer, if possible? We deliberately hide `job_config.checkpoint` inside Checkpointer and let the trainer always call checkpointer API. This usage looks okay but would be good to hide if possible.

**Review by tianyu-l — APPROVED** [review]
LGTM, nice refactor!
